dts: entropy: cleanup Kconfig and replace CONFIG_ENTROPY_NAME with DT_ENTROPY_NAME

galak opened the pull request.

mbolivar-nordic committed:
include: devicetree.h: API for /chosen zephyr,foo

Add a devicetree/zephyr.h header, which is meant to contain
definitions for /chosen properties specific to Zephyr.

Currently, this just deals with zephyr,entropy.  We add a
DT_CHOSEN_ZEPHYR_ENTROPY_LABEL macro which expands to the label for the
node pointed to by zephyr,entropy.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: entropy: Remove Kconfig HAS_DTS_ENTROPY

Now that all entropy drivers use DTS we can remove HAS_DTS_ENTROPY being
set everywhere as well as Kconfig ENTROPY_NAME since that is now coming
from DT_ENTROPY_NAME.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
drivers: entropy: replace CONFIG_ENTROPY_NAME with DT macro

Replace CONFIG_ENTROPY_NAME with DT_CHOSEN_ZEPHYR_ENTROPY_LABEL.  We now
set zephyr,entropy in the chosen node of the device tree to the entropy
device.

This allows us to remove CONFIG_ENTROPY_NAME from dts_fixup.h.  Also
remove any other stale ENTROPY related defines in dts_fixup.h files.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

carlescufi reviewed (approved):
LGTM

MaureenHelm commented on soc/arm/atmel_sam/same70/dts_fixup.h:
```diff
@@ -127,7 +127,7 @@
 #define DT_USBHS_RAM_BASE_ADDRESS	DT_ATMEL_SAM_USBHS_40038000_RAM_0
 
-#define CONFIG_ENTROPY_NAME			DT_ATMEL_SAM_TRNG_40070000_LABEL
+#define DT_ENTROPY_NAME			DT_ATMEL_SAM_TRNG_40070000_LABEL
```
Why isn't this `DT_LABEL(DT_INST(0, atmel_sam_trng))`?

galak replied:
just historic, leaving what that part of the define alone.

MaureenHelm commented on soc/arm/nxp_imx/rt/dts_fixup.h:
```diff
@@ -25,7 +25,7 @@
 #define DT_IRQ_ETH1_IEEE1588_TMR		DT_NXP_KINETIS_PTP_402D4000_PTP_IRQ_0
 
-#define CONFIG_ENTROPY_NAME			DT_NXP_KINETIS_TRNG_400CC000_LABEL
+#define DT_ENTROPY_NAME			DT_NXP_KINETIS_TRNG_400CC000_LABEL
```
Why isn't this `DT_LABEL(DT_INST(0, nxp_kinetis_trng))`

galak replied:
just historic, leaving what that part of the define alone.

anangl commented:
@galak A related change in the hal_nordic module is also needed (see https://github.com/zephyrproject-rtos/hal_nordic/pull/32).

galak committed:
devicetree.h: Move device includes to the end

Move the include of files from include/devicetree/*.h to the end so that
when those files are processes they have access to all previous defined
macros.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

MaureenHelm reviewed (approved):
needs rebase

pabigot commented on doc/zephyr.doxyfile.in:
```diff
@@ -1996,7 +1996,8 @@ PREDEFINED             = "CONFIG_ARCH_HAS_CUSTOM_BUSY_WAIT" \
                          "__attribute__(x)=" \
-                         "__syscall="
+                         "__syscall=" \
+                         "DT_DOXYGEN"
```
unrelated, undocumented, change?  What's this for?  (Is this commit from some other PR?  @mbolivar-nordic)

Later: Looks like it's used to generate documentation in Doxygen, but IMO this should not be named to suggest it's restricted to devicetree usage.  There are many other cases where conent that is conditionally included in a header should be included or excluded for Doxygen-generated material.  In other projects I've done this in `doxygen.cfg.in`:
```
# Use "#if defined(@PROJECT_UC@_DOXYGEN)" when providing a stub declaration
# to which documentation is attached.
# Use "/* @cond DOXYGEN_EXCLUDE */ ... /* @endcond */" around the real
# declarations.
PREDEFINED += "@PROJECT_UC@_DOXYGEN"
```

galak replied:
This was pull in from PR #23891 to get the commits from that PR related to devicetree/zephyr.h

pabigot replied:
OK.  Now I see this:
```
 * Getting doxygen to play along with all the dts-specific ifdeffery
￼ * proved too complex for DT_CHOSEN_ZEPHYR_FLASH_LABEL, so we document
￼ * everything under a DT_DOXYGEN define.
```
in the point of use, so I guess it's explained.  I still think a generic "only emit this when running under Doxygen" solution is more useful, but that's probably more hassle to get through.

pabigot reviewed (commented):
This seems to make some sense, in spite of it possibly starting a slide down a slope to a massive `chosen` section.

galak commented:
Reworked this PR based on discussion in dev-review.  Pulled in some commits from @mbolivar-nordic PR on FLASH.  This now uses a chosen property to specify the entropy device.

galak commented on include/devicetree/zephyr.h:
```diff
@@ -45,6 +45,8 @@
 #define DT_CHOSEN_ZEPHYR_FLASH_LABEL DT_LABEL(DT_CHOSEN(zephyr_flash))
 #endif
 
+#define DT_CHOSEN_ZEPHYR_ENTROPY_LABEL DT_LABEL(DT_CHOSEN(zephyr_entropy))
```
I think DT_NODE_HAS_PROP covers both if the node exists and if the property exists.

galak replied:
What part should be under DT_DOXYGEN?

mbolivar-nordic replied:
> I think DT_NODE_HAS_PROP covers both if the node exists and if the property exists.

Right you are. OK then.

> What part should be under DT_DOXYGEN?

Never mind, I see it now.
